### PR TITLE
ARROW-7608: [C++][Dataset] Add the ability to list files in FileSystemSource

### PR DESCRIPTION
**Pull request by fsaintjacques**
It also enables instrospection in R of dataset classes.

### CODE REVIEW

File: r/R/dataset.R [hunk 1]
```diff
@@ -163,30 +169,59 @@ DatasetFactory$create <- function(sources) {
...
+    #' @description
+    #' Return the files contained in this `Source`
+    files = function() dataset___FSSource__files(self),
+    #' @description
+    #' Return the format of files in this `Source`
+    format = function() FileFormat$create(dataset___FSSource__format(self))
   )
 )
 
+Source$create <- function(x) {
```
Comment (nealrichardson) [inline r/R/dataset.R]: We don't use the `$create` methods for taking pointers and wrapping them in their class, generally. `create` is for R user facing input. Maybe the `..dispatch` pattern is a better fit.

File: r/R/dataset.R [hunk 2]
```diff
@@ -163,30 +169,59 @@ DatasetFactory$create <- function(sources) {
...
       shared_ptr(Schema, dataset___Source__schema(self))
-    }
+    },
+    #' @description
+    #' Return the Source's type.
+    type = function() dataset___Source__type_name(self)
+  )
+)
+
+#' @name FileSystemSource
```
Comment (nealrichardson) [inline r/R/dataset.R]: Add a note in the `Source` docs above that FSS is a type of Source. Also should list the methods/active bindings there (the `@description` doclet isn't being used yet).

File: r/tests/testthat/test-dataset.R [hunk 3]
```diff
@@ -347,9 +353,11 @@ test_that("Assembling a Dataset manually and getting a Table", {
 })
 
 test_that("Assembling multiple SourceFactories with DatasetFactory", {
-  src1 <- open_source(file.path(dataset_dir, 1), format = "parquet")
+  dir1 <-file.path(dataset_dir, 1)
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R]:
```suggestion
  dir1 <- file.path(dataset_dir, 1)
```

File: r/tests/testthat/test-dataset.R [hunk 4]
```diff
@@ -347,9 +353,11 @@ test_that("Assembling a Dataset manually and getting a Table", {
...
+  src1 <- open_source(dir1, format = "parquet")
   expect_is(src1, "FileSystemSourceFactory")
-  src2 <- open_source(file.path(dataset_dir, 2), format = "parquet")
+  dir2 <-file.path(dataset_dir, 2)
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R]:
```suggestion
  dir2 <- file.path(dataset_dir, 2)
```

**Review by nealrichardson — COMMENTED** [review]
Thanks, this looks helpful. I want to think some more about how we handle the pattern in R of determining which subclass of object to instantiate, I'm not fully satisfied. 

We'll also want to enhance the R Dataset print method, or provide some other details method that exposes this information in a natural readable way. We could defer but it might be useful to do in this PR in case there are other methods we need to add in C++.

**Comment by fsaintjacques** [comment]
The OSX failure is a Flight unrelated issue.

File: r/R/dataset.R (hunk 1 above)
Comment (fsaintjacques) [reply r/R/dataset.R]: Moved to dispatch.

**Comment by jorisvandenbossche** [comment]
I don't know how that information is exactly stored, but: would it also be possible to have a way to get the partition expression for each file?

**Comment by fsaintjacques** [comment]
Absolutely.

**Comment by kszucs** [comment]
Perhaps `(partition_expression, file)` iterator would be better to expose under `files()`

```python
for partition, path in source.files():
    pass
```

Either way we should have a jira about it.

File: r/tests/testthat/test-dataset.R [hunk 5]
```diff
@@ -383,15 +383,21 @@ test_that("Assembling a Dataset manually and getting a Table", {
...
 
   src <- factory$Finish(schm)
-  expect_is(src, "Source")
+  expect_is(src, "FileSystemSource")
   expect_is(src$schema, "Schema")
+  expect_is(src$format, "ParquetFileFormat")
   expect_equal(names(schm), names(src$schema))
 
+
+  expect_equivalent(src$files, c(file1, file2))
```
Comment (pitrou) [inline r/tests/testthat/test-dataset.R]: Is it guaranteed that files are returned in order?

**Comment by bkietz** [comment]
There's a windows path formatting issue: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/30694260/job/8rs3y80x55fclkrk#L5502
I'll take a look

File: r/tests/testthat/test-dataset.R (hunk 5 above)
Comment (bkietz) [reply r/tests/testthat/test-dataset.R]: Yes: they will always be in a depth first visitation order.

**Review by nealrichardson — APPROVED** [review]
I reworked the subclass dispatching in R and am now satisfied with this for now. I'll want to wire up prettier print methods using this info, but I can do that in a followup.

**Comment by fsaintjacques** [comment]
Merci Neal!

**Comment by jorisvandenbossche** [comment]
Is this ready to be merged?

**Comment by fsaintjacques** [comment]
it isn't due to the windows path issue, I'm not sure if I should dig into it or just remove the R test for now (it'll still work, just the path is funky).

**Comment by nealrichardson** [comment]
I've pushed a commit to try to make the tests in R work on windows (basically, tell `file.path()` to use `/` as path separator so that the paths will compare as equal. If that doesn't work and we don't want to lose more time on it, we can add `skip_on_os("windows")` before the tests and move on.

**Comment by bkietz** [comment]
The odd path segments are introduced by `tempfile()` (before the failing test runs) so I think the solution is to wrap it with `normalizePath(tempfile(), winslash = "/"))`

**Pull request closed without merging**